keps/sig-scheduling/3521-pod-schedule-readiness/README.md

```diff
@@ -0,0 +1,989 @@
+
+Both in this section and below, follow the guidelines of the [documentation
+style guide]. In particular, wrap lines to a reasonable length, to make it
+easier for reviewers to cite specific portions, and to minimize diff churn on
+updates.
+
+[documentation style guide]: https://github.com/kubernetes/community/blob/master/contributors/guide/style-guide.md
+-->
+
+This KEP aims to add a `.spec.schedulePaused` field to Pod's API, to mark a Pod's schedule readiness.
+Integrators can mutate this field to instruct scheduler whether it's ready for scheduling or not.
+
+## Motivation
+
+<!--
+This section is for explicitly listing the motivation, goals, and non-goals of
+this KEP.  Describe why the change is important and the benefits to users. The
+motivation section can optionally provide links to [experience reports] to
+demonstrate the interest in a KEP within the wider Kubernetes community.
+
+[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
+-->
+
+Pods are naturally perceived ready for scheduling once created. Kubernetes scheduler does
+its due diligence to find nodes to place all pending Pods. However, in a real-world case,
+some Pods may stay in a "miss-essential-resources" state for a long period. These pods,
+actually churn the scheduler (and downstream integrators like Cluster AutoScaler) in an
+unnecessary manner.
+
+Lacking the knob to flag Pods as schedule-ready/unready wastes scheduling cycles on retrying
+Pods that are determined to be unschedulable. As a result, it delays the scheduling of other Pods,
+and also lowers the overall scheduling throughput. Moreover, it imposes restrictions to vendors to
+develop some in-house features (such as hierarchical quota) natively.
+
+This proposal describes APIs and mechanics to allow users/controllers to control when a pod is
+ready to be considered for scheduling.
+
+### Goals
+
+<!--
+List the specific goals of the KEP. What is it trying to achieve? How will we
+know that this has succeeded?
+-->
+
+- Define API to mark Pod as schedule-paused/unpaused.
+- Design a new Enqueue extension point to customize Pod's queueing behavior.
+- A default enqueue plugin to honor the new API semantics.
+
+### Non-Goals
+
+<!--
+What is out of scope for this KEP? Listing non-goals helps to focus discussion
+and make progress.
+-->
+- Update the Pod's condition if it carries `.spec.schedulePaused=true`.
+- Focus on in-house use-cases of the Enqueue extension point.
+
+## Proposal
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation. What is the desired outcome and how do we measure success?.
+The "Design Details" section below is for the real
+nitty-gritty.
+-->
+
+We propose a new field `.spec.schedulePaused` to the Pod API. The field is defaulted to `false`.
```

Comment:
I was thinking (and I suggested it to @ahg-g offline just now) that it might be worth making this a list of strings, in a similar fashion to finalizers.

Once the list is empty, the pod is allowed to be scheduled.

This allows multiple controllers/webhooks to decide whether a pod can be scheduled based on different criteria.

It might be useful to have an endpoint just for this field (and disallow edition via general patch/update) so that there are less chances that a controller accidentally removes all the strings. Or maybe we can just  have a simple update validation: only one string can be removed at a time.

Although, for finalizers we don't do that. The only related thing is an RBAC rule that allows controllers to just modify finalizers but nothing else.

Reply:
@pwschuurman this should help simplify https://github.com/kubernetes/enhancements/pull/3336

Reply:
Thanks @alculquicondor for this idea. This looks more extensible, and we may name it
`ScheduleReadinessGates []string`.

I'm interested to learn from API experts if there are any traps in adopting this finalizers-like API pattern.

Reply:
It's great, then each controller can have its own enqueue plugin if it needs.

Reply:
Actually, the whole point of the list is that developers don't have to implement scheduler plugins. They just control when to remove their own string from the list.

Reply:
random thoughts from finalizers:
* is ordering important or can any gate in the list be removed at any time?
* is it important to protect against one gate owner messing with a gate belonging to another other (either accidentally or intentionally)? finalizers don't enforce any additional permissions by default (the OwnerReferencesPermissionEnforcement admission plugin that enforces this is optional and off by default), so any API client with update permission can add/remove any finalizer
* what is the expectation for gate owners to report status when blocked? is there any concept of a timeout?

Reply:
Re "ordering": ordering is not important. Gates can be removed after the pod gets created, uni-directionally. (i.e., you cannot append/remove back and forth).
Re "permission": each gate owner is _supposed_ to be a good citizen and only remove their gate(s) when needed. Permission protection is a good-to-have feature, but I don't see it as a blocker at least for alpha.
Re "timeout": I guess it's the same story with finalizers. If some gate owners get disconnected, then the gates cannot be cleared and thus stay in that state. There is no timeout concept on the scheduler side, so totally relies on gate owners to clear the gates.

Reply:
> Permission protection is a good-to-have feature, but I don't see it as a blocker at least for alpha.

If you don't have it initially, I'm *really* skeptical it will ever get added. That's what happened with finalizers, and it's why there's no compatible way to walk back that feature's wide-open posture

Reply:
Gotcha. We have to make it right in the very beginning if we go with this approach.

Reply:
+1 to gate it from the beginning if there are the mechanics to do it.

Reply:
I assume we will need to setup a endpoint and the path would include the string to be removed/added so that we can setup RBAC on a per string basis. Is this feasible @liggitt ?

Reply:
@ahg-g I'm not sure RBAC supports a so-called "dynamic" endpoint/path.

IIUC, we may need to come up with something like `ScheduleReadinessGates []OwnerRefs`, each ownerRef needs to be GVRed, so that RBAC can tell if the operation is allowed.

Reply:
Don't we support something similar for namespace access?

Reply:
It turns out that a couple of other KEPs are also stuck at this same decision point and there is a discussion going on regarding the path forward [1]. To unblock this KEP, Jordan is suggesting to "make a generic reference to subdividing permission/access with a yet-to-be-determined mechanism consistent with liens and finalizers". 

[1] https://groups.google.com/g/kubernetes-sig-auth/c/U7rQiOEsnmA/m/aU82M1aKCgAJ?utm_medium=email&utm_source=footer

Reply:
at first glance, these seem similar to custom pod readiness gates, where a list of conditions can be registered in the pod spec, and until all those conditions are present/true, the pod is not considered ready

would a parallel list of "scheduler gates" make sense here? they would be registered in `spec.schedulerGates`, the external actors would update pod status to add their conditions, and once they were all present/true, the scheduler would fire

that has the benefit of:
1. matching custom readiness gates approach
2. giving an existing status field / mechanism for actors to publish their individual condition
3. giving an existing status field for those actors to communicate updates / negative status (register their condition with a value of False and a reason, etc)

Reply:
whether it is a list of readiness gates or a boolean field I am concerned with defaulting to false which changes the current behavior where pods are considered ready to schedule by default.

Also mutation rules need to be spec-ed. For the same of the next part let us assume it is a a boolean field. Here are the rules:

1. `false` => `true`: OK, Pod was waiting on something that was availed.
2. `true` => `false`: NOT OK, Pod was ready and thus may have been already somewhere in the scheduling, starting process and can not be brought be back `waiting for something that was availed`

Reply:
@khenidak you're understanding right actually. It's just the field is named "schedulingPaused", so defaulted to false, indicating it's scheduling ready by default. The rules you mentioned are also correct, just swap `false` and `true`.

Reply:
one more thing `paused` typically means ready for `play` and `pause` again. Should consider `.spec.schedulerReady` or something that does not implicitly imply that the state machine can move backward and forward.

Reply:
@liggitt if we go with `spec.schedulerGates`, I suppose the following problems are hard (or impossible) to tackle?

1. we cannot gate which user can control which schedulerGate as any condition can be updated as long as you have update/patch permission on pod/status?
2. we cannot easily enforce a uni-directional transition for a particular ready condition?

Reply:
I think another way to ask Jordan’s question is about the future:

“Do we envision a workload controlled workflow for scheduling that would require coordination between multiple actors in the future”

If we can’t say yes, the boolean is appropriate.  If we *could* say yes, we should enumerate what our arguments for or against those briefly here.

Reply:
I’m a little hesitant to endorse a workflow conceptual field without very strong and immediate use cases that justify it.  We have always tried to find the minimal declarative structure with maximum composability - even simple lists of strings are rarely sufficient for end-user (vs system) control.  So the bar would be very high to add such a field now.

Reply:
> “Do we envision a workload controlled workflow for scheduling that would require coordination between multiple actors in the future”

I don't think so. Each gate is supposed to represent a particular condition, and behave statelessly.

Reply:
> So the bar would be very high to add such a field now.

@smarterclayton to clarify, by "such a field" you meant the `[]string` field brainstormed in this thread; instead of the originally proposed `bool` field, right?

Reply:
I'm in favor of @liggitt's idea

Reply:
> Do we envision a workload controlled workflow for scheduling that would require coordination between multiple actors in the future

Actually, yes.

@pwschuurman mentioned a use case related to migrations. The original motivation for this KEP was about queueing. What if users wanted to have both behaviors? The different controllers gating a pod to be scheduled need to coordinate.

Reply:
Thinking about this last night, one of the advantages of a readiness gate approach is that it allows a human to create pods that won't be scheduled yet (through either an unknown gate or an explicitly predefined gate).  So that's two use cases worth calling out (including the stateful set orchestrator):

1. A workload orchestrator can use a scheduler readiness gate to perform custom control of when pods are released to scheduling (statefulset use case)
2. A human can create a pod and delay scheduling until such a time as other resources are initialized iteratively, or allow it

Both of those bring up key design considerations:

1. How should a workload controller handle pods that remain in scheduling unready due to paused for a long-time?  Do we need to do anything additional over "unable to schedule"?  Should the scheduler have a time based window before a paused pod is flagged?
2. Can a cluster be brought down by large numbers of unscheduled pods?  We have a component responsible for lifecycling completed pods (pod GC), but not one for unscheduled pods.  Are there additions we need to make to quota or alerting to help administrators manage this list?  (apologies if this is in KEP and I missed it)

Reply:
Very good points on the considerations.

Re 1: for now, I don't quite consider we should impose a time based mechanics to dishonor the API (paused state), even conditionally. The paused state is supposed to be a strong scheduling gate until it's cleared.
Re 2: it does make sense, esp. in operation's view, to prevent this feature from being misused - maybe due to faulty controllers. But I'm hesitant to add support from the beginning for the hyperthetical requirement. I'd like to document this potential behavior for now, and revisit once we get some real feedback. Maybe associated with a PDB-like mechanics.

I will update the KEP to describe the potential symptoms.

Reply:
Regarding 1: I also don't think we should impose time limits for two reasons:
1) currently we don't have time limits for pods that fail to schedule
2) I don't think this feature will encourage creating more pods and so it shouldn't really impact scalability in a negative way; in fact, this feature enables a new design pattern that should improve scalability: currently the scheduler continuously attempts to find a node for unschedulable pods, and so wasting scheduling cycles; this feature gives orchestrators a mechanism to tell the scheduler to ignore those pods until a later time, and so practically saving scheduling cycles.

Regarding 2, we can add a quota scope to limit the number of "not-ready to schedule pods": https://kubernetes.io/docs/concepts/policy/resource-quotas/#quota-scopes

Reply:
> currently we don't have time limits for pods that fail to schedule

I should have been more clear with "time limit", I was asking two questions together (the first you answered), the second being

How long before we set the status condition on a paused pod with a reason for why the pod is not scheduled?  An old client may not even be able to observe the paused field (for instance, a web interface like octant or a cli that doesn't use server side get), and so to those users a paused pod is indistinguishable from "scheduler has hung / is unavailable".  

Re: 2, I think you are correct that this feature is net neutral on number of pods (actually, slightly reduces the aggregate number of pods) because it create no more pods than would normally get scheduled.  If someone attempted to implement reservation via creating a pause pod they will have to update or delete it and recreate it to use the reservation.

Reply:
> How long before we set the status condition on a paused pod with a reason

Scheduling-wise, we can separate the conditions into two categories:

- [PodScheduled](https://github.com/kubernetes/kubernetes/blob/8f269d6df2a57544b73d5ca35e04451373ef334c/staging/src/k8s.io/api/core/v1/types.go#L2656) - for paused pods, as mentioned in Non-Goals section we won't update this condition, as it stands for a "scheduler tried but failed", and it's used by CA heavily to determine whether the pod needs a new node to accormordate.
- to-be-determined condition type like "PodSchedulingPaused" - I'm not quite sure if we need it as it seems this is for lower-version clients only. My understanding is even with this new condition, the client (UI/cli) still cannot understand it without updating their own logic, then given code changes are needed anyway, is there a big difference with upgrading to a new client-go version so the paused field is aware?

Reply:
Right, the exiting `PodScheduled` condition carries the meaning that the scheduler attempted to find a node and failed, which currently triggers autoscaling in cluster-autoscaler. There are two things we can do here:

1) CA should start to recognize the flag
2) We could use the same condition, but use a different `reason` since `reason` is supposed to be machine readable. This will also require a change in CA, but it is a change that we can patch back to older CA versions (for example, the change would be to ignore the pod if the `PodScheduled` condition is false and reason is "SchedulingNotReady")

Reply:
> we can patch back to older CA versions

Maybe we just make changes to the latest CA version as CA [suggests](https://github.com/kubernetes/autoscaler/tree/master/cluster-autoscaler#releases): "We don't do cross version testing or compatibility testing in other environments."

Reply:
If we're worried about CA and others reacting to `PodScheduled=false`, we *could* set `PodScheduled` to `Unknown` after some time with an appropriate reason, and then the server side printer for pods could leverage that if we had to.  I think the hard part would be trying to decide how long to wait (15s? 30s?), and again, whether it is worth it.

It may be simpler to update the server side printer for pods to output "waiting" when scheduling is paused (in which case the condition is less critical).

Reply:
The original plan was to not set the status; instead, exposing the Pod's state by events, metrics and CLI support.

That was based on not breaking lower versions of CA. However, if CA is guaranteed to only support the exact matched k8s version, update a pod's status as `PodScheduled=paused` doesn't sound like a bad idea. In that case, we don't need to rely on a timeout - internally the paused pod will still go through a lightweight scheduling process (enqueue extension point) where we can update the status.

Reply:
We could also have creation default the pod Scheduled condition to false and reason `WaitingForGate`.

Reply:
> We could also have creation default the pod Scheduled condition to false and reason `WaitingForGate`.

Make sense to me, this saves us a mutating request compared to having the scheduler doing it. But I am not sure if we should use condition status `unknown` or `false`. I prefer `false` if we can guarantee compatibility with CA,  @x13n what do you think?

Reply:
> We could also have creation default the pod Scheduled condition to false and reason WaitingForGate.

That's fair. Just to clarify: the condition "PodScheduled=WaitingForGate" is applied when it goes through the lightweight scheduling process, instead of defaulted by `pkg/apis/core/<something>`.

Reply:
No, we can default it at pod creation.

Reply:
seems we don't have precedence of defaulting a condition? (I checked pkg/registry/core/pod/strategy.go)

But, if API folks are fine with the pattern that at least save the scheduler's efforts.

Reply:
We're adding a Kube feature, defaulting can do many things, while it's novel, it's certainly worth trying.  If we come up with a reason to remove it before beta, that's low cost, but we do have defaulting of status (namespace status phase defaulting).  I think it's a nice usability win, and we can record "will look for implications of novel condition defaulting for GA".  We may only be able to do the defaulting during creation (can't do it during update), so that's a possible wrinkle.  Could simply list it as a run-at for mitigation.

Reply:
> I think it's a nice usability win

+1. Already added it in the monitoring section.